### ISSUE TITLE
ZFS init fails on Centos due to not loaded module

### ISSUE THREAD

**Issue by cf-sewe** [bodgit/puppet-zfs#30]
I would like to get your opinion about following issue. 
When the Puppet ZFS initializes on CentOS 7.7 (fresh system), it fails with zfs-import-scan due to not loaded zfs kernel module.

Do you have an idea how I can do the ZFS initial install in a way to workaround this issue? I see on Ubuntu or Centos 6 you are explicitly loading the kernel module. I guess ZFS module will only be loaded automatically on Centos7 if a ZFS volume already exists on the server?

What I wonder is, if this might be a new issue related to Centos 7.7 update, new ZFS 0.8.3 behaviour or new version of zfs puppet module. Because previously, the kernel module was loaded automatically and the zfs puppet init completed fine without manual intervention.

So this part already fails:
`
    class { '::zfs':
      kmod_type   => 'dkms',
      zfs_arc_max => to_bytes('16 G'),
      zfs_arc_min => to_bytes('1 G'),
    }
`

With following error "The ZFS modules are not loaded":

> 2020-02-03 15:38:13 +0100 /Stage[main]/Zfs::Install/Package[kernel-devel]/ensure (notice): created
> 2020-02-03 15:38:14 +0100 /Stage[main]/Fail2ban::Service/Service[fail2ban]/ensure (notice): ensure changed 'stopped' to 'running' (corrective)
> 2020-02-03 15:39:41 +0100 /Stage[main]/Zfs::Install/Package[zfs]/ensure (notice): created
> 2020-02-03 15:39:41 +0100 Puppet (err): Systemd start for zfs-import-scan failed!
> journalctl log for zfs-import-scan:
> -- Logs begin at Mon 2020-02-03 15:37:44 CET, end at Mon 2020-02-03 15:39:41 CET. --
> Feb 03 15:39:41 servername systemd[1]: Starting Import ZFS pools by device scanning...
> Feb 03 15:39:41 servername zpool[28269]: The ZFS modules are not loaded.
> Feb 03 15:39:41 servername zpool[28269]: Try running '/sbin/modprobe zfs' as root to load them.
> Feb 03 15:39:41 servername systemd[1]: zfs-import-scan.service: main process exited, code=exited, status=1/FAILURE
> Feb 03 15:39:41 servername systemd[1]: Failed to start Import ZFS pools by device scanning.
> Feb 03 15:39:41 servername systemd[1]: Unit zfs-import-scan.service entered failed state.
> Feb 03 15:39:41 servername systemd[1]: zfs-import-scan.service failed.
> 2020-02-03 15:39:41 +0100 /Stage[main]/Zfs::Service/Service[zfs-import-scan]/ensure (err): change from 'stopped' to 'running' failed: Systemd start for zfs-import-scan failed!
> journalctl log for zfs-import-scan:
> -- Logs begin at Mon 2020-02-03 15:37:44 CET, end at Mon 2020-02-03 15:39:41 CET. --
> Feb 03 15:39:41 servername systemd[1]: Starting Import ZFS pools by device scanning...
> Feb 03 15:39:41 servername zpool[28269]: The ZFS modules are not loaded.
> Feb 03 15:39:41 servername zpool[28269]: Try running '/sbin/modprobe zfs' as root to load them.
> Feb 03 15:39:41 servername systemd[1]: zfs-import-scan.service: main process exited, code=exited, status=1/FAILURE
> Feb 03 15:39:41 servername systemd[1]: Failed to start Import ZFS pools by device scanning.
> Feb 03 15:39:41 servername systemd[1]: Unit zfs-import-scan.service entered failed state.
> Feb 03 15:39:41 servername systemd[1]: zfs-import-scan.service failed.
> 2020-02-03 15:39:41 +0100 /Stage[main]/Zfs::Service/Service[zfs-mount] (notice): Dependency Service[zfs-import-scan] has failures: true
> 2020-02-03 15:39:41 +0100 /Stage[main]/Zfs::Service/Service[zfs-mount] (warning): Skipping because of failed dependencies
> 2020-02-03 15:39:41 +0100 /Stage[main]/Zfs::Service/Service[zfs-share] (warning): Skipping because of failed dependencies
> 2020-02-03 15:39:41 +0100 /Stage[main]/Profile::Redhat/Exec[zfs-bootstrap.sh] (warning): Skipping because of failed dependencies
> 2020-02-03 15:39:41 +0100 /Stage[main]/Profile::Redhat/Exec[zfs-swap.sh] (warning): Skipping because of failed dependencies

System: CentOS 7.7.1908
ZFS Version installed by bodgit::zfs: zfs-dkms-0.8.3-1.el7.noarch
ZFS module: mod 'bodgit-zfs', '2.2.7'

**Comment by bodgit** [comment]
It should always load the module, at least it always has done. The modprobe hack is there because the old SysV init scripts all check whether the module is loaded and bail out with an exit code of 0 which then just confuses Puppet as it thinks the service started successfully.

I'm fairly sure the Systemd unit files had a modprobe line in them, something like `ExecStartPre=-/sbin/modprobe zfs` which is basically doing the same thing. Perhaps that's been dropped in the latest versions, I'm away from my ZFS host so I can't check until later, maybe you can confirm if they still contain that?

I assume if you manually run `modprobe zfs` it does actually load the module?

**Comment by cf-sewe** [comment]
Yes, they (ZOL) replaced the modprobe with another logic: in 0.8.3 https://github.com/zfsonlinux/zfs/commit/a6cc97566c54255c98612c801f2c9d9be2d0e8f8#diff-fe9214b8af4a902509a60f7290d138a7 They say with that change a explicit modprobe is no longer needed, because calling the zfs binary will auto-load the module on first request. However this didnt work, at least on my system, maybe because no ZFS pool exists initially..

A manual `modprobe zfs` works fine, but this is a manual action which is needed after ZFS is installed, and should be ideally done by the zfs puppet module to avoid failed puppet runs and manual interaction.

Probably adding a explicit modprobe zfs also for RHEL7/8 systems wont hurt, would be great if you could consider adding this.

**Comment by bodgit** [comment]
Thanks for the pointer to the commit, it does indeed look like they've removed all of the modprobe lines. I think it should still load properly even if you didn't have any ZFS pools created. I'm not sure why it's not loading automatically, but if it works after a reboot then something somewhere needs restarting, my hunch would be udev.

**Comment by bodgit** [comment]
I've just pushed some changes to hopefully fix this. Basically I've re-added the `ExecStartPre=` lines back to the systemd units by way of some override files. I needed to do it this way as the Debian 10 package is a broken in much the same way and these need to be present before the package is installed for the first time.

I've also added EL8 support now that the EPEL module has been updated. If you're able to make use of a git checkout the changes are all in master, otherwise I'll cut a 2.2.8 release shortly.
